tutorials: add a tutorial for loggert and message (#261)
co-log/co-log: tutorials/3-loggert-with-message/README.md

# Message and LoggerT

> The previous tutorial shows how to use `loggert` with simple message.

Reviewer: ```suggestion
The previous tutorial shows how to use `LoggerT` with simple message.
```

```

## Preamble: imports and language extensions

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}

module Main (main) where

import Colog (
    LogAction,
    Message,
    Msg (..),
    WithLog,
    cmap,
    fmtMessage,
    formatWith,
    logError,
    logInfo,
    logTextStderr,
    logTextStdout,
    msgSeverity,
    msgText,
    richMessageAction,
    showSeverity,
    usingLoggerT,
 )

import Data.Text (Text)
import Prelude hiding (log)
```

## LoggerT

The `LoggerT` monad transformer wraps a ReaderT that keeps `LogAction` in its context. It denotes a
computation of logging. So you define your monadic actions with the `WithLog` constraint that allows you to perform logging.
The `Message` allows you to log with severity.

````haskell
example1 :: WithLog env Message m => m ()
example1 = do
    logInfo "this is a demo log for message!"

example2 :: WithLog env Message m => m ()
example2 = do
    logError "Mayday! Mayday! Here is an example of error log\n"
```

> The `WithLog` constraint has three type parameters: the application environment,
> the type of the message and the monad. The actions constraint by `WithLog` could be used as `LoggerT`.

Reviewer: ```suggestion
The `WithLog` constraint has three type parameters:
 * `env` – application environment
 * `msg` – type of the message
 * `m` – monad 
 
 The actions constraint by `WithLog` could be used as `LoggerT`.
```

## Message

> The message is parametrized by the Severity type through `Msg`, which is a general logging message data type.

Reviewer: ```suggestion
The message is parametrized by the `Severity` type through `Msg`, which is a general logging message data type.
```

```idris
type Message = Msg Severity
data Msg sev = Msg
    { msgSeverity :: !sev
    , msgStack    :: !CallStack
    , msgText     ::  Text
    }

> ````

Reviewer: ```suggestion
```
```

> Hence, besides printing the callstack as the `loggert`` with simple message tutorial shows, messages support print severity and thread id as well.

Reviewer: ```suggestion
Hence, besides printing the callstack, as the `LoggerT` with simple message tutorial shows, this example supports printing severity and thread id as well.
```
